src/openlcb/VirtualMemorySpace.hxx

```diff
@@ -0,0 +1,329 @@
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * \file VirtualMemorySpace.hxx
+ *
+ * Implementation of a memory space where the values are not stored in a
+ * contiguous storage area but are read and written via callbacks.
+ *
+ * @author Balazs Racz
+ * @date 10 Aug 2020
+ */
+
+#ifndef _OPENLCB_VIRTUALMEMORYSPACE_HXX
+#define _OPENLCB_VIRTUALMEMORYSPACE_HXX
+
+#include "openlcb/ConfigEntry.hxx"
+#include "openlcb/MemoryConfig.hxx"
+#include "utils/SortedListMap.hxx"
+
+namespace openlcb
+{
+
+/// Implementation of a memory space where the values are not stored in a
+/// contiguous storage area but are read and written via callbacks.
+class VirtualMemorySpace : public MemorySpace
+{
+public:
+    /// @returns whether the memory space does not accept writes.
+    bool read_only() override
+    {
+        return isReadOnly_;
+    }
+    /// @returns the lowest address that's valid for this block.
+    address_t min_address() override
+    {
+        return minAddress_;
+    }
+    /// @returns the largest valid address for this block.  A read of 1 from
+    /// this address should succeed in returning the last byte.
+    address_t max_address() override
+    {
+        return maxAddress_;
+    }
+
+    /// @return the number of bytes successfully written (before hitting end
+    /// of space).
+    /// @param destination address to write to
+    /// @param data to write
+    /// @param len how many bytes to write
+    /// @param error if set to non-null, then the operation has failed. If the
+    /// operation needs to be continued, then sets error to
+    /// MemorySpace::ERROR_AGAIN, and calls the Notifiable
+    /// @param again when a re-try makes sense. The caller should call write
+    /// once more, with the offset adjusted with the previously returned
+    /// bytes.
+    size_t write(address_t destination, const uint8_t *data, size_t len,
+        errorcode_t *error, Notifiable *again) override
+    {
+        if (destination > maxAddress_ || destination + len <= minAddress_)
```

Comment:
Should parentheses be around (destination + len)? I'm unsure of the C rules and I have always used parentheses in this case.

Reply:
done